router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/H2FailureAccrualFactory.scala

```diff
@@ -0,0 +1,276 @@
+            // should not mark dead. We probe again in an attempt to
+            // resolve this ambiguity, but we could also mark dead for a
+            // fixed period of time, or even mark alive.
+            startProbing()
+          case None =>
+        }
+      case _ =>
+    }
+  }
+
+  private[this] val onServiceAcquisitionFailure: Throwable => Unit = self.synchronized { _ =>
+    stopProbing()
+    didFail()
+  }
+
+  protected def didSucceed(): Unit = self.synchronized {
+    // Only count revivals when the probe succeeds.
+    state match {
+      case ProbeClosed =>
+        revivalCounter.incr()
+        policy.revived()
+        state = Alive
+      case _ =>
+    }
+    policy.recordSuccess()
+  }
+
+  private[this] def markDeadFor(duration: Duration) = self.synchronized {
+    // In order to have symmetry with the revival counter, don't count removals
+    // when probing fails.
+    if (state == Alive) removalCounter.incr()
+
+    state = Dead
+
+    val timerTask = timer.schedule(duration.fromNow) { startProbing() }
+    reviveTimerTask = Some(timerTask)
+
+    removedForCounter.incr(duration.inMilliseconds.toInt)
+
+    didMarkDead()
+  }
+
+  /**
+   * Called by FailureAccrualFactory after marking an endpoint dead. Override
+   * this method to perform additional actions.
+   */
+  protected def didMarkDead() = {}
+
+  /**
+   * Enter 'Probing' state.
+   * The service must satisfy one request before accepting more.
+   */
+  protected def startProbing() = self.synchronized {
+    state = ProbeOpen
+    cancelReviveTimerTask()
+  }
+
+  /**
+   * Exit 'Probing' state (if necessary)
+   *
+   * The result of the subsequent request will determine whether the factory transitions to
+   * Alive (successful) or Dead (unsuccessful).
+   */
+  private[this] def stopProbing() = self.synchronized {
+    state match {
+      case ProbeOpen =>
+        probesCounter.incr()
+        state = ProbeClosed
+      case _ =>
+    }
+  }
+
+  private[this] def classifyResponse(classifier: H2Classifier, h2ReqRep: H2ReqRep): Try[Response] =
+    classifier.responseClassifier.lift(h2ReqRep) match {
+      case Some(ResponseClass.Successful(_)) =>
+        didSucceed()
+        h2ReqRep.response
+      case Some(ResponseClass.Failed(_)) =>
+        didFail()
+        h2ReqRep.response
+      case None =>
+        h2ReqRep.response match {
+          case Return(rep) =>
+            if (rep.stream.isEmpty) {
+              classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, None))))
+              Return(rep)
+            } else {
+              val stream = rep.stream.onFrame {
+                case Return(f) if f.isEnd =>
+                  classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, Some(Return(f))))))
+                case Throw(e) =>
+                  classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, Some(Throw(e))))))
```

Comment:
wondering if there's an easy way to collapse these match arms...probably there isn't.